icefall/score_estimator.py

```diff
@@ -0,0 +1,188 @@
+        super().__init__()
+        self.embedding = nn.Linear(
+            in_features=input_dim,
+            out_features=hidden_dim
+        )
+        self.output = nn.Linear(
+            in_features=hidden_dim,
+            out_features=2
+        )
+        self.sigmod = nn.Sigmoid()
+
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        x = self.embedding(x)
+        x = self.sigmod(x)
+        x = self.output(x)
+        mean, var = x[:, 0], x[:, 1]
+        var = torch.exp(var)
+        return mean, var
+
+
+def get_parser():
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter
+    )
+
+    parser.add_argument(
+        "--input-dim",
+        type=int,
+        default=5,
+        help="Dim of input feature.",
+    )
+
+    parser.add_argument(
+        "--hidden-dim",
+        type=int,
+        default=20,
+        help="Neural number of didden layer.",
+    )
+
+    parser.add_argument(
+        "--batch-size",
+        type=int,
+        default=10,
+        help="Batch size of dataloader.",
+    )
+
+    parser.add_argument(
+        "--epoch",
+        type=int,
+        default=20,
+        help="Training epochs",
+    )
+
+    parser.add_argument(
+        "--learning-rate",
+        type=float,
+        default=1e-4,
+        help="Learning rate.",
+    )
+
+    parser.add_argument(
+        "--exp_dir",
+        type=Path,
+        default=Path("conformer_ctc/exp"),
+        help="Directory to store experiment data.",
+    )
+    return parser
+
+
+def main():
+    parser = get_parser()
+    args = parser.parse_args()
+    torch.manual_seed(42)
+    torch.cuda.manual_seed(42)
+
+    setup_logger(f"{args.exp_dir}/rescore/log")
+
+    model = ScoreEstimator(
+        input_dim = args.input_dim,
+        hidden_dim = args.hidden_dim
+    )
+
+    model = model.to("cuda")
+    optimizer = torch.optim.Adam(model.parameters(), lr=args.learning_rate)
+    loss_fn = nn.GaussianNLLLoss()
+
+    train_dataloader = DataLoader(
+        Dataset(f"{args.exp_dir}/rescore/feat", "train"),
+        collate_fn=DatasetCollateFunc(),
+        batch_size=args.batch_size,
+        shuffle=True
+    )
+    dev_dataloader = DataLoader(
+        Dataset(f"{args.exp_dir}/rescore/feat", "dev"),
+        collate_fn=DatasetCollateFunc(),
+        batch_size=args.batch_size,
+        shuffle=True
+    )
+
+    for epoch in range(args.epoch):
+        model.train()
+        training_loss = 0.0
+        step = 0
+        for x, y in train_dataloader:
+            mean, var = model(x.cuda())
+            loss = loss_fn(mean, y, var)
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+            training_loss += loss.item()
+            step += len(y)
+        training_loss /= step
+
+        dev_loss = 0.0
+        step = 0
+        model.eval()
+        for x, y in dev_dataloader:
+            mean, var = model(x.cuda())
```

Comment:
I would recommend you to define a variable `device`, which can be either a CPU or a CUDA device, and use `model(x.to(device))`.